File: src/blanker.ts
import type { Comment, Node } from "oxc-parser";
import BlankString from "./blank-string.js";
import { Trivia } from "./trivia.js";
import {
    type OnError,
    VISIT_BLANKED,
    VISIT_JS,
    type VisitResult,
} from "./types.js";

/**
 * Per-run state plus the primitive blanking operations. The walk logic lives in
 * walk.ts and friends; they coordinate through an instance of this class.
 */
export class Blanker {
    readonly output: BlankString;
    readonly trivia: Trivia;

    /** True while the previously emitted JS did not end with a `;`. */
    semicolonNeeded = false;
    /** Statement currently being walked, used by the `as`/`satisfies` rule. */
    parentStatement: Node | undefined;
    /** Set by walk.ts; the recursive dispatch over AST nodes. */
    visitNode: (node: Node) => VisitResult = () => VISIT_JS;

    constructor(
        readonly src: string,
        comments: readonly Comment[],
        private readonly onError: OnError | undefined,
    ) {
        this.output = new BlankString(src);
        this.trivia = new Trivia(src, comments);
    }

    /** Report an unsupported construct; its source stays in the output. */
    report(node: Node): void {
        this.onError?.({ type: node.type, start: node.start, end: node.end });
    }

    blankRange(start: number, end: number): void {
        this.output.blank(start, end);
    }

    /** Blank [node.start, node.end). */
    blankExact(node: Node): void {
        this.output.blank(node.start, node.end);
    }

    /**
     * Blank a statement-like node that is being fully erased. A leading `;` is
     * emitted when the previous emitted JS lacks one, so that a following
     * statement cannot merge into it (ASI protection).
     */
    blankStatement(node: Node): void {
        if (this.semicolonNeeded) {
            this.output.blankButStartWithSemi(node.start, node.end);
        } else {
            this.output.blank(node.start, node.end);
        }
    }

    /** Blank a `: T` type annotation; oxc spans include the colon. */
    blankTypeAnnotation(annotation: Node): void {
        this.output.blank(annotation.start, annotation.end);
    }

    /** Blank a node and, when directly followed by a comma, the comma too. */
    blankExactAndOptionalTrailingComma(node: Node): void {
        const comma = this.trivia.scanChar(node.end);
        this.output.blank(
            node.start,
            this.src[comma] === "," ? comma + 1 : node.end,
        );
    }

    /**
     * Whether the node text ends with `;` or a same-line `;` directly follows
     * it. Parity helper: the TypeScript AST counts trailing semicolons as part
     * of statement spans, oxc does not.
     */
    endsWithSemicolon(node: Node): boolean {
        return this.trivia.endsWithSemicolon(node.start, node.end);
    }

    /** Visit a nested node, keeping the semicolon state up to date. */
    visitNested(node: Node): VisitResult {
        const result = this.visitNode(node);
        if (result === VISIT_JS) {
            this.semicolonNeeded = !this.endsWithSemicolon(node);
        }
        return result;
    }

    /**
     * Visit an array of children. When `isStatementLike` is set, each child is
     * tracked as `parentStatement` (used by the `as`/`satisfies` rule) and
     * updates the semicolon state after each visited child. Entering a function
     * body resets the semicolon state, mirroring the nested execution context.
     */
    visitNodeArray(
        nodes: readonly Node[],
        isStatementLike: boolean,
        isFunctionBody: boolean,
        visitChild: (node: Node) => VisitResult,
    ): VisitResult {
        const previousParentStatement = this.parentStatement;
        const previousSemicolonNeeded = this.semicolonNeeded;
        if (isFunctionBody) {
            this.semicolonNeeded = false;
        }
        for (const node of nodes) {
            if (isStatementLike) {
                this.parentStatement = node;
            }
            if (visitChild(node) === VISIT_JS) {
                this.semicolonNeeded = !this.endsWithSemicolon(node);
            }
        }
        this.parentStatement = previousParentStatement;
        if (isFunctionBody) {
            this.semicolonNeeded = previousSemicolonNeeded;
        }
        return this.semicolonNeeded ? VISIT_JS : VISIT_BLANKED;
    }
}

File: native/src/blanker.rs
//! Port of `src/core/blanker.ts` — per-run state plus the primitive blanking
//! operations. The walk logic lives in `walk.rs` and friends; they coordinate
//! through an instance of this struct.

use oxc_parser::{Kind, Token};
use oxc_span::Span;

use crate::blank_string::BlankString;
use crate::trivia::TokenIndex;

/// Information about a TypeScript-only construct that has runtime semantics
/// and therefore cannot be erased. Positions are byte offsets into the input.
#[derive(Debug, Clone)]
pub struct UnsupportedSyntax {
    pub node_type: &'static str,
    pub start: u32,
    pub end: u32,
}

pub struct Blanker<'a> {
    pub output: BlankString,
    pub tokens: TokenIndex<'a>,
    /// True while the previously emitted JS did not end with a `;`.
    pub semicolon_needed: bool,
    /// Unsupported constructs reported so far.
    pub reports: Vec<UnsupportedSyntax>,
}

impl<'a> Blanker<'a> {
    pub fn new(src: &'a str, tokens: &'a [Token]) -> Self {
        Self {
            output: BlankString::default(),
            tokens: TokenIndex::new(src, tokens),
            semicolon_needed: false,
            reports: Vec::new(),
        }
    }

    /// Report an unsupported construct; its source stays in the output.
    pub fn report(&mut self, node_type: &'static str, span: Span) {
        self.reports.push(UnsupportedSyntax {
            node_type,
            start: span.start,
            end: span.end,
        });
    }

    pub fn blank_range(&mut self, start: u32, end: u32) {
        self.output.blank(start, end);
    }

    /// Blank [span.start, span.end).
    pub fn blank_span(&mut self, span: Span) {
        self.output.blank(span.start, span.end);
    }

    /// Blank [span.start, span.end).
    pub fn blank_exact(&mut self, span: Span) {
        self.output.blank(span.start, span.end);
    }

    /// Blank a statement-like node that is being fully erased. A leading `;` is
    /// emitted when the previous emitted JS lacks one, so that a following
    /// statement cannot merge into it (ASI protection).
    pub fn blank_statement(&mut self, span: Span) {
        if self.semicolon_needed {
            self.output.blank_but_start_with_semi(span.start, span.end);
        } else {
            self.output.blank(span.start, span.end);
        }
    }

    /// Blank a `: T` type annotation; oxc spans include the colon.
    pub fn blank_type_annotation(&mut self, span: Span) {
        self.output.blank(span.start, span.end);
    }

    /// Blank a node and, when directly followed by a comma, the comma too.
    pub fn blank_exact_and_optional_trailing_comma(&mut self, span: Span) {
        let end = match self.tokens.token_from(span.end) {
            Some(token) if token.kind() == Kind::Comma => token.span().end,
            _ => span.end,
        };
        self.output.blank(span.start, end);
    }

    /// Whether the node text ends with `;` or a same-line `;` directly follows
    /// it.
    pub fn ends_with_semicolon(&self, span: Span) -> bool {
        self.tokens.ends_with_semicolon(span.start, span.end)
    }

    /// Erase the marker token (`?` or `!`) sitting directly before `anchor`.
    pub fn blank_marker_char(&mut self, anchor: u32, marker: Kind) {
        if let Some(span) = self.tokens.marker_before(anchor, marker) {
            self.output.blank(span.start, span.end);
        }
    }
}
